qmail-qfilter.1
.TH qmail-qfilter 1 2020-12-25
.SH NAME
qmail-qfilter \- filter messages before they reach the queue
.SH SYNOPSIS
.B qmail-qfilter
.I filter
[
.I -- filter ...
]
.SH DESCRIPTION
.B qmail-qfilter
provides a Unix-filter interface for programs to
inspect and act on messages before they reach the queue.
Each
.I filter
can
accept,
modify before accepting,
reject temporarily, or
reject permanently.
.P
.B qmail-qfilter
runs the given sequence of filters (separated by
.BR -- )
before invoking
.BR qmail-queue .
If any filter rejects, the message is rejected.
.P
Each
.I filter
receives the message on its standard input
and the envelope on file descriptor 3.
To pass the message unmodified,
.I filter
may simply exit 0;
otherwise, it should write the modified message to its standard output
and/or the modified envelope to file descriptor 4.
(These are temporary files managed by 
.BR qmail-qfilter .)
.P
If every
.I filter
has accepted the message,
.B qmail-qfilter
executes
.BR qmail-queue .
Exceptions:
.TP 5
.BR 1 .
If
.B QQF_QMAILQUEUE
is set, it specifies the alternate program to be executed. See
.BR "ENVIRONMENT VARIABLES" .
.TP
.BR 2 .
If one or more filters have written to file descriptor 5
(another temporary file managed by
.BR qmail-qfilter ),
its contents specify the alternate program to be executed.
.P
In any case, the program being executed must adhere to the
.B qmail-queue
interface.
.SH "EXIT CODES"
Following the
.B qmail-queue
interface,
.I filter
can exit
.TP 5
.B 0
to accept the message,
.TP 
.B 31
to reject the message permanently, or
.TP
.B 82
to reject the message (temporarily or permanently) with a custom error string.
.P
.I filter
can additionally exit
.TP 5
.B 99
to have
.B qmail-qfilter
exit 0 immediately without running any further filters.
.P
If
.I filter
exits anything other than 0 or 99,
.B qmail-qfilter
exits with
.IR filter's
exit code.
.P
Exception:
If
.B qmail-qfilter
can't create temporary files or execute filters, it exits
.TP 5
.B 51
out of memory,
.TP
.B 53
write error, or
.TP
.B 81
internal error.
.P
Exception:
If
.B qmail-qfilter
can't read or parse envelope data, it exits
.TP 5
.B 91
bad envelope data.
.P
Otherwise,
.B qmail-qfilter exits with
.BR qmail-queue 's
exit code.
.SH "ENVIRONMENT VARIABLES"
Each
.IR filter 's
environment contains:
.TP 5
.B QMAILPPID
The PID of
.BR qmail-qfilter 's
parent process (e.g.,
.BR qmail-smtpd ).
.TP
.B QMAILUSER
The user part of the envelope sender.
.TP
.B QMAILHOST
The host part of the envelope sender.
.TP
.B QMAILRCPTS
The list of envelope recipients, each followed by a newline.
.TP
.B ENVSIZE
The size of the envelope.
.TP
.B MSGSIZE
The length of the message.
.TP
.B NUMRCPTS
The number of recipients.
.P
For compatibility with previous versions,
.B QMAILNAME
is unset.
.P
Other programs included with qmail which invoke
.B qmail-queue
will invoke the contents of
.B QMAILQUEUE
instead, if set.
.B qmail-qfilter
does not.
This avoids recursive loops with configurations that invoke
.B qmail-qfilter
by setting
.BR QMAILQUEUE .
To override
.BR qmail-qfilter 's
invocation of
.BR qmail-queue ,
set
.B QQF_QMAILQUEUE
instead.
.SH EXAMPLES
To add SpamAssassin headers to every message:
.P
.EX
    spamc
.EE
.P
To reject messages with long
.B Date:
headers:
.P
.EX
    #!/usr/bin/perl
    while (<>) {
      print;
      last if /^\n$/o;
      exit 31 if /^Date: .{80,}/oi;
    }
    while (<>) {
      print;
    }
.EE
.SH WARNINGS
If
.I filter
is
.BR "qmail-inject -n",
you may want to run it with
.BR MAILUSER ,
.BR USER ,
and
.B LOGNAME
unset, e.g.:
.P
.B "env -u MAILUSER -u USER -u LOGNAME qmail-inject -n"
.P
A message with an excessive number of recipients (more than 64K bytes of
recipient data on Linux) will cause filter programs to fail to execute
and the message to be rejected.
.P
Each
.I filter
gets file descriptor 5 open to the same temporary file.
When writing a program name to it, make sure to write a trailing ASCII NUL byte.
Otherwise, multiple filters could overwrite the value in undesirable ways.
.SH HISTORY
.B qmail-qfilter
was originally written by Bruce Guenter.
.SH "SEE ALSO"
qmail-queue(8)


CHANGES
20200520 remove wrappers for obsolete mail clients without maildir support
20200520 version: notqmail 1.08.
20200519 bug: fix several other places where variables could overflow.
20200519 security: fix signedness wraparound in substdio_{put,bput}().
         (CVE-2005-1515)
20200519 security: fix possible signed integer overflow in commands().
         (CVE-2005-1514)
20200519 security: fix integer overflow in stralloc_readyplus().
         (CVE-2005-1513)
20200514 bug: qmail-pop3d runs as root if root authenticates.
         impact: vector for dictionary attack on root password.
         fix: exit 1, same as a failed checkpassword login.
20200514 code: changed qmail-popup interface (leave kid's fd 2 alone).
20200511 code: remove tryshsgr.c and use uid_t, gid_t types.
20200511 add first unittest, run with "make test"
20191227 fixed possible overflow in alloc() with allocations close to 4GiB
20191226 add missing return types to main()
20190919 fix ipme with multiple addresses on the same interface on platforms
         without sa_len
20190904 code: use stdint.h to get the definition for uint32
20190903 code: do not parse header recipients for "qmail-inject -a"
20190826 code: if $QMAILREMOTE is set, it's invoked instead of qmail-remote.
20190822 cleanup: remove unused dnsmxip.
20190822 cleanup: remove unused dnscname.
20190819 version: notqmail 1.07.
20190810 code: remove incorrect usage of vfork()
20190710 code: add instchown to set ownership and mode on installed files.
20190709 code: add instpackage to install without root access.
20190714 code: removed idedit, install-big. var-qmail packages are no more.
20190708 code: lookup uid/gid at runtime.
20190711 portability problem: explicitly initialize strerr_sys and
         define BIND_8_COMPAT to work around Mac OS X.
20190709 doc: rename INSTALL, SENDMAIL to INSTALL.md, SENDMAIL.md to
         support building on case-insensitive filesystems.
20190715 code: detect and prefer utmpx where available.
20190708 code: use DESTDIR environment variable as root directory in install.
20071130 version: netqmail 1.06
20071130 legal: qmail-1.03 is now in the public domain
20051103 doc: dot-qmail.9 updated for changed (19980613) conf-patrn default, tnx ADM
20040121 version: netqmail 1.05
20040121 code: qmail-smtpd is protected from exceedingly long (eg 2GB)
         header lines
20040121 code: qmail_lspawn, qmail-newmrh, qmail-newu, and qmail-rspawn
         are protected from misbehaving on hosts where the size of an
         integer is not the same as the size of a character pointer
         (eg 64 bit hosts with 32 bit ints)
20031027 version: netqmail 1.04.
20031027 doc: INSTALL points to http://lifewithqmail.org/lwq.html
20031027 doc: qmail.7 identifies installation as netqmail and points to
         http://qmail.org/
20031027 doc: qmail-queue.8 adds explanation of $QMAILQUEUE
20031027 doc: qmail-log.5 adds reference to errors from $QMAILQUEUE script
20031027 doc: FAQ also points to http://cr.yp.to/qmail/faq.html and
         http://qmail.org/
20031027 code: qmail-smtpd identifies itself as netqmail
20031027 code: if $QMAILQUEUE is set, it's invoked instead of qmail-queue
20031024 code: changed errno from int to #include.
20031024 code: fixed .qmail parsing bug (accessing an array out of bounds)
         that was discovered by Erik Sjölund <erik.sjolund@gmail.com>).
20031024 code: recognize 0.0.0.0 as a local address.
20031024 code: sendmail's -f flag now overrides environment variables.
19980615 version: qmail 1.03.
19980614 doc: eliminated BIN.setup in favor of a web page.
19980614 code: added other auto* to qmail-showctl output.
19980614 doc: added pointer to immhf.html in qmail-header.5.
19980614 doc: added note to TEST.receive about SMTP command format.
19980614 doc: added FAQ 5.6 on qmail-qmqpd.
19980614 code: removed unused variables in idedit.c.
19980613 code: changed conf-patrn to 002.
19980613 doc: moved SENDMAIL lower in INSTALL.
19980612 code: added install-big.
19980612 code: added BIN.Makefile.
19980612 doc: added BIN.README, BIN.setup.
19980612 code: switched to new install.
19980611 code: added idedit.
19980611 doc: added FAQ 1.3 on $QMAILMFTFILE.
19980611 doc: used bouncesaying in FAQ 5.5.
19980611 code: added except.
19980611 code: added bouncesaying.
19980611 code: allowed unbracketed IP addresses in dns_ipplus() and
         dns_mxip().
19980611 code: allowed spaces after colon in non-bracketed addresses in
         qmail-smtpd.
19980611 doc: cleaned up UPGRADE.
19980528 bug: qmail-smtpd skips first character in rcpthosts() call.
         tnx NND. impact: qmail-smtpd crashes on empty address; and it
	 allows relaying to ""@any.host. fix: use addr.s.
19980515 doc: expanded flock discussion in INSTALL.mbox.
19980515 doc: eliminated flock warning from INSTALL.maildir.
19980515 doc: split REMOVE.binmail out of INSTALL.
19980515 doc: split REMOVE.sendmail out of INSTALL.
19980515 doc: split TEST.deliver and TEST.receive out of INSTALL and
         UPGRADE.
19980515 doc: integrated INSTALL.boot into INSTALL.
19980515 code: cleaned up final output in qmail-qmqpd.c.
19980514 doc: updated procmail notes in INSTALL.mbox. tnx JRM.
19980514 doc: changed FAQ 4.4 to point to INSTALL.mbox for procmail.
         tnx JRM.
19980514 code: separated HELO and EHLO; single-line response for HELO.
         tnx to various people.
19980430 version: qmail 1.02.
19980430 doc: updated SECURITY.
19980430 doc: fixed FAQ 4.9. tnx KB.
19980430 code: changed quote2() to avoid quoting <>.
19980429 code: changed quote_need() to quote empty local parts. tnx HHO.
19980428 doc: added status notes to INSTALL and UPGRADE.
19980428 code: skip setting environment in sendmail.c if PROTO is set.
19980428 code: eliminated recipientmap.
19980428 code: added virtual users to qmail-send.c. tnx RN.
19980428 code: eliminated domain from rewrite() in qmail-send.c.
19980428 code: added binm1, binm1+df, binm2, binm2+df, binm3, binm3+df.

qmail-qfilter.c
/* Copyright (C) 2001,2004-2005 Bruce Guenter <bruceg@em.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "qmail-qfilter.h"

#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include "alloc.h"
#include "byte.h"
#include "env.h"
#include "fd.h"
#include "fmt.h"
#include "fork.h"
#include "str.h"
#include "substdio.h"

static const char* binqqargs[2];
static size_t envelope_len = 0;
static size_t message_len = 0;

static char errbuf[SUBSTDIO_OUTSIZE];
static substdio sserr = SUBSTDIO_FDBUF(write,2,errbuf,sizeof(errbuf));
static char pidstr[FMT_ULONG];

static void qqf_debug(const char *s1,const char *s2)
{
  if (!env_get("QQF_DEBUG")) return;

  substdio_puts(&sserr,"qmail-qfilter pid ");
  substdio_puts(&sserr,pidstr);
  substdio_puts(&sserr," ");
  substdio_puts(&sserr,s1);
  substdio_puts(&sserr," ");
  substdio_puts(&sserr,s2);
  substdio_puts(&sserr,"\n");
  substdio_flush(&sserr);
}

static void qqf_debug_ulong(const char *s1,unsigned long val)
{
  char strnum[FMT_ULONG];

  byte_zero(strnum,FMT_ULONG);
  fmt_ulong(strnum,val);
  return qqf_debug(s1,strnum);
}

static void die(int exitcode)
{
  qqf_debug_ulong("exit",exitcode);
  exit(exitcode);
}

static void die_nomem(void)    { die(51); }
static void die_write(void)    { die(53); }
static void die_internal(void) { die(81); }
static void die_envelope(void) { die(91); }

static int env_put2_ulong(const char* key, unsigned long val)
{
  char strnum[FMT_ULONG];

  byte_zero(strnum,FMT_ULONG);
  fmt_ulong(strnum,val);
  return env_put2(key,strnum);
}

static size_t parse_sender(const char* envelope)
{
  const char* ptr = envelope;
  size_t pos;
  char* at;
  size_t len = str_len(envelope);

  if (*ptr != 'F') die_envelope();
  ++ptr;

  env_unset("QMAILNAME");

  if (!*ptr) {
    if (!env_put("QMAILUSER=")) die_nomem();
    if (!env_put("QMAILHOST=")) die_nomem();
    return 2;
  }

  pos = str_rchr(ptr, '@');
  len = str_len(ptr);
  if (pos >= len)
    at = 0;
  else
    at = (char *)ptr + pos;

  if (at) {
    char *user;
    size_t user_len;

    len = str_len(at);

    user_len = str_len(ptr) - len;
    user = malloc(user_len + 1);
    for (int i = 0; i < user_len; i++)
      user[i] = ptr[i];
    user[user_len] = 0;

    if (!env_put2("QMAILUSER",user)) die_nomem();
    if (!env_put2("QMAILHOST",at+1)) die_nomem();
    ptr = at;
  }
  else {
    if (!env_put2("QMAILUSER",ptr)) die_nomem();
    if (!env_put("QMAILHOST=")) die_nomem();
  }

  return ptr + len + 1 - envelope;
}

static void parse_recipients(const char* envelope, int offset)
{
  size_t len = envelope_len - offset;
  const char* ptr = envelope + offset;
  char* buf = malloc(len);
  char* tmp = buf;
  unsigned long count = 0;

  while (ptr < envelope + envelope_len && *ptr == 'T') {
    size_t rcptlen = str_len(++ptr);

    memcpy(tmp, ptr, rcptlen);
    tmp[rcptlen] = '\n';
    tmp += rcptlen + 1;
    ptr += rcptlen + 1;
    ++count;
  }
  *tmp = 0;
  if (!env_put2("QMAILRCPTS",buf)) die_nomem();
  if (!env_put2_ulong("NUMRCPTS", count)) die_nomem();
  free(buf);
}

static void parse_envelope(int fd)
{
  const char* envelope;
  size_t offset;

  envelope = mmap(0, envelope_len, PROT_READ, MAP_PRIVATE, fd, 0);
  if (envelope == MAP_FAILED) die_nomem();
  offset = parse_sender(envelope);
  parse_recipients(envelope, offset);
  munmap((void*)envelope, envelope_len);
}

static int invisible_readwrite_tempfile()
{
  char filename[sizeof(TMPDIR)+19] = TMPDIR "/fixheaders.XXXXXX";
  int fd = mkstemp(filename);

  if (fd == -1) die_write();
  if (unlink(filename) == -1) die_write();

  return fd;
}

static void rewind_fd(int fd)
{
  if (lseek(fd, 0, SEEK_SET) != 0) die_write();
}

static size_t copy_fd_contents_and_close(int fdin, int fdout)
{
  size_t bytes;
  int tmp = invisible_readwrite_tempfile();
  char buf[BUFSIZE];
  substdio ssin = SUBSTDIO_FDBUF(read,fdin,buf,sizeof(buf));
  substdio sstmp = SUBSTDIO_FDBUF(write,tmp,buf,sizeof(buf));

  for (bytes = 0;;) {
    ssize_t r = substdio_get(&ssin,buf,sizeof(buf));

    if (r == -1) die_write();
    if (r == 0)
      break;

    if (substdio_putflush(&sstmp,buf,sizeof(buf)) == -1) die_write();
    bytes += r;
  }

  close(fdin);
  rewind_fd(tmp);
  if (fd_move(fdout,tmp) == -1) die_write();

  return bytes;
}

command* parse_args_to_linked_list_of_filters(int argc, char* argv[])
{
  command* tail = 0;
  command* head = 0;

  while (argc > 0) {
    command* cmd;
    int end = 0;

    while (end < argc && str_diff(argv[end], "--"))
      ++end;
    if (end == 0) die_internal();
    argv[end] = 0;
    cmd = malloc(sizeof(command));
    cmd->argv = argv;
    cmd->next = 0;
    if (tail)
      tail->next = cmd;
    else
      head = cmd;
    tail = cmd;
    ++end;
    argv += end;
    argc -= end;
  }

  return head;
}

static void invisible_readwrite_tempfileat(int fd)
{
  int tmp;

  close(fd);
  tmp = invisible_readwrite_tempfile();
  if (fd_move(fd,tmp) == -1) die_write();
}

static void move_unless_empty(int src, int dst, const void* reopen,
                              size_t* var)
{
  struct stat st;

  if (fstat(src, &st) != 0) die_internal();
  if (st.st_size > 0) {
    if (fd_move(dst,src) == -1) die_write();
    *var = st.st_size;
    if (reopen) {
      invisible_readwrite_tempfileat(src);
      if (src == ENVELOPE_OUT)
        parse_envelope(ENVELOPE_IN);
    }
  }
  else
    if (!reopen)
      close(src);
  rewind_fd(dst);
}

static char *qq_overridden_by_filter(int fd)
{
  struct stat st;
  char* buf = 0;

  if (fstat(fd, &st) != 0) die_internal();
  if (st.st_size > 0) {
    buf = malloc(st.st_size + 1);
    if (!buf) die_internal();
    if (read(fd, buf, st.st_size) != st.st_size) die_internal();
    buf[st.st_size] = 0;
  }
  close(fd);

  return buf;
}

void run_filters_in_sequence(const command* first)
{
  const command* c;

  invisible_readwrite_tempfileat(MESSAGE_OUT);
  invisible_readwrite_tempfileat(ENVELOPE_OUT);

  for (c = first; c; c = c->next) {
    pid_t pid;
    int status;

    if (!env_put2_ulong("ENVSIZE", envelope_len)) die_nomem();
    if (!env_put2_ulong("MSGSIZE", message_len)) die_nomem();
    pid = fork();
    if (pid == -1) die_nomem();
    if (pid == 0) {
      execvp(c->argv[0], c->argv);
      die_internal();
    }
    qqf_debug_ulong(c->argv[0],pid);
    if (waitpid(pid, &status, WUNTRACED) == -1) die_internal();
    if (!WIFEXITED(status)) die_internal();
    if (WEXITSTATUS(status))
      die((WEXITSTATUS(status) == QQ_DROP_MESSAGE) ? 0 : WEXITSTATUS(status));
    move_unless_empty(MESSAGE_OUT, MESSAGE_IN, c->next, &message_len);
    move_unless_empty(ENVELOPE_OUT, ENVELOPE_IN, c->next, &envelope_len);
    rewind_fd(QMAILQUEUE_OVERRIDE);
  }
}

void setup_qqargs(int fd)
{
  if (!binqqargs[0])
    binqqargs[0] = qq_overridden_by_filter(fd);
  if (!binqqargs[0])
    binqqargs[0] = env_get("QQF_QMAILQUEUE");
  if (!binqqargs[0])
    binqqargs[0] = "bin/qmail-queue";
}

void exec_qq(void)
{
  if (fd_move(1,ENVELOPE_IN) == -1) die_write();
  qqf_debug("execv",binqqargs[0]);
  execv(binqqargs[0], (char**)binqqargs);
  die_internal();
}

void prepare_to_run_filters(void)
{
  fmt_ulong(pidstr,getpid());
  if (!env_put2_ulong("QMAILPPID", getppid())) die_nomem();
  qqf_debug("ppid",env_get("QMAILPPID"));

  message_len = copy_fd_contents_and_close(0, 0);
  envelope_len = copy_fd_contents_and_close(1, ENVELOPE_IN);
  parse_envelope(ENVELOPE_IN);
  invisible_readwrite_tempfileat(QMAILQUEUE_OVERRIDE);
}
